client/go/outline/iptable/packet_listener.go

```diff
@@ -0,0 +1,319 @@
+	defaultListener transport.PacketListener
+	defaultConn     net.PacketConn
+
+	listenerTable IPTable[transport.PacketListener]
+	connMap       map[netip.Addr]net.PacketConn
+	connMapLock   sync.Mutex
+
+	forwardingContext            context.Context
+	closePacketForwardingContext context.CancelFunc
+	forwardedPackets             chan incomingPacket
+	forwardCounter               sync.WaitGroup
+
+	lastRemainingPacket     incomingPacket
+	lastRemainingPacketLock sync.Mutex
+}
+
+var _ net.PacketConn = (*packetConn)(nil)
+
+func newPacketConn(
+	parentContext context.Context,
+	listenerTable IPTable[transport.PacketListener],
+	defaultListener transport.PacketListener,
+) (*packetConn, error) {
+	forwardingContext, closeForwardingContext := context.WithCancel(parentContext)
+
+	var defaultConn net.PacketConn
+	if defaultListener != nil {
+		connAttempt, err := defaultListener.ListenPacket(forwardingContext)
+
+		if err != nil {
+			defaultConn = nil
+		} else {
+			defaultConn = connAttempt
+		}
+	}
+
+	conn := &packetConn{
+		closePacketForwardingContext: closeForwardingContext,
+		connMap:                      make(map[netip.Addr]net.PacketConn),
+		connMapLock:                  sync.Mutex{},
+		defaultConn:                  defaultConn,
+		defaultListener:              defaultListener,
+		forwardedPackets:             make(chan incomingPacket, packetQueueSize),
+		forwardingContext:            forwardingContext,
+		listenerTable:                listenerTable,
+		lastRemainingPacket:          incomingPacket{},
+		lastRemainingPacketLock:      sync.Mutex{},
+	}
+
+	if conn.defaultConn != nil {
+		conn.forwardPackets(conn.defaultConn)
+	}
+
+	return conn, nil
+}
+
+func (conn *packetConn) ReadFrom(result []byte) (n int, addr net.Addr, err error) {
+	if len(conn.lastRemainingPacket.data) > 0 {
+		conn.lastRemainingPacketLock.Lock()
+		numBytes := copy(result, conn.lastRemainingPacket.data)
+		conn.lastRemainingPacket.data = conn.lastRemainingPacket.data[numBytes:]
+		conn.lastRemainingPacketLock.Unlock()
+
+		return numBytes, conn.lastRemainingPacket.addr, nil
+	}
+
+	packet, ok := <-conn.forwardedPackets
+	if !ok {
+		return 0, nil, net.ErrClosed
+	}
+
+	numBytes := copy(result, packet.data)
+
+	conn.lastRemainingPacketLock.Lock()
+	conn.lastRemainingPacket.data = packet.data[numBytes:]
+	conn.lastRemainingPacket.addr = packet.addr
+	conn.lastRemainingPacketLock.Unlock()
+
+	return numBytes, packet.addr, nil
+}
+
+func (conn *packetConn) WriteTo(packet []byte, addr net.Addr) (numBytes int, err error) {
+	var ip netip.Addr
+	if udpAddr, ok := addr.(*net.UDPAddr); ok {
+		if parsedIP, ok := netip.AddrFromSlice(udpAddr.IP); ok {
+			ip = parsedIP
+		}
+	}
+	if !ip.IsValid() {
+		return 0, fmt.Errorf("could not parse valid IP from address %v (%T)", addr, addr)
+	}
+
+	ip = ip.Unmap()
+
+	conn.connMapLock.Lock()
+	subconn := conn.connMap[ip]
+	conn.connMapLock.Unlock()
+
+	if subconn == nil {
+		listener, ok := conn.listenerTable.Lookup(ip)
```

Comment:
Is `IPTable` thread safe? Do we allow users to modify the table after creation?

Reply:
Not a bad idea. Or we can have a immutable IPTable, and any modifications will create a new immutable IPTable.